src/main/java/io/cdap/plugin/SFTPCopyAction.java

```diff
@@ -129,65 +131,81 @@ public String getVariableNameHoldingFileList() {
     FileSystem fileSystem = FileSystem.get(conf);
-
     destination = fileSystem.makeQualified(destination);
     if (!fileSystem.exists(destination)) {
       fileSystem.mkdirs(destination);
     }
-
-    try (SFTPConnector SFTPConnector = new SFTPConnector(config.getHost(), config.getPort(), config.getUserName(),
-                                                      config.getPassword(), config.getSSHProperties())) {
-      ChannelSftp channelSftp = SFTPConnector.getSftpChannel();
-
-      Vector files = channelSftp.ls(config.getSrcDirectory());
-
-      List<String> filesCopied = new ArrayList<>();
-      for (int index = 0; index < files.size(); index++) {
-        Object obj = files.elementAt(index);
-        if (!(obj instanceof ChannelSftp.LsEntry)) {
-          continue;
+    SFTPConnector sftpConnector = null;
+    if (config.getAuthTypeBeingUsed().equals(SFTPConstants.PRIVATE_KEY_SELECT)) {
+      try { sftpConnector = new SFTPConnector(config.getHost(), config.getPort(),
+        config.getUserName(), config.getPrivateKey(), config.getPassphrase(), config.getSSHProperties());
+        copySFTPFiles(fileSystem, destination, sftpConnector, context);
+      } catch (Exception e){
+        throw new RuntimeException(String.format("Error occurred while connecting to SFTP Server %s", e.getMessage()));
+      } finally {
+        if (sftpConnector != null) {
+          sftpConnector.close();
         }
-        ChannelSftp.LsEntry entry = (ChannelSftp.LsEntry) obj;
-        if (".".equals(entry.getFilename()) || "..".equals(entry.getFilename())) {
-          // ignore "." and ".." files
-          continue;
+      }
+    } else {
+      try { sftpConnector = new SFTPConnector(config.getHost(), config.getPort(),
+        config.getUserName(), config.getPassword(), config.getSSHProperties());
+        copySFTPFiles(fileSystem, destination, sftpConnector, context);
+      } catch (Exception e) {
+        throw new RuntimeException(String.format("Error occurred while connecting to SFTP Server %s", e.getMessage()));
+      } finally {
+        if (sftpConnector != null) {
+          sftpConnector.close();
         }
+      }
+    }
+  }
 
-        // Ignore files that don't match the given file regex
-        if (!Strings.isNullOrEmpty(config.fileNameRegex)) {
-          String fileName = entry.getFilename();
-          if (!fileName.matches(config.fileNameRegex)) {
-            LOG.debug("Skipping file {} since it doesn't match the regex.", fileName);
-            continue;
-          }
+  private void copySFTPFiles(FileSystem fileSystem, Path destination, SFTPConnector SFTPConnector,
+                            ActionContext context) throws SftpException, IOException {
+    ChannelSftp channelSftp = SFTPConnector.getSftpChannel();
+    Vector files = channelSftp.ls(config.getSrcDirectory());
+    List<String> filesCopied = new ArrayList<>();
+    for (int index = 0; index < files.size(); index++) {
+      Object obj = files.elementAt(index);
```

Comment:
can this throw index out of bounds exception?

Reply:
Im not sure. I havent seen this in my testing, but I can add indexoutofbounds to the thrown exceptions

Reply:
the files vector is empty in case source directory does not contain any values?

Reply:
I'm not sure I can test this though. I'm pretty sure it errors and fails the pipeline currently saying the file location is empty.

Reply:
What happens when `getSrcDirectory` is empty?

Reply:
if SrcDirectory is empty no error occurs just no files are processed and the logs include no files moved since files being moved are logged to debug. If the dir does not exist the current exception is thrown in the catch with the added message of no such file